<a href="https://colab.research.google.com/github/msaipraveenkumar/EDA_Assg_NYC_Taxi_Starter/blob/main/EDA_Assg_NYC_Taxi_Starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **New York City Yellow Taxi Data**

## Objective
In this case study you will be learning exploratory data analysis (EDA) with the help of a dataset on yellow taxi rides in New York City. This will enable you to understand why EDA is an important step in the process of data science and machine learning.

## **Problem Statement**
As an analyst at an upcoming taxi operation in NYC, you are tasked to use the 2023 taxi trip data to uncover insights that could help optimise taxi operations. The goal is to analyse patterns in the data that can inform strategic decisions to improve service efficiency, maximise revenue, and enhance passenger experience.

## Tasks
You need to perform the following steps for successfully completing this assignment:
1. Data Loading
2. Data Cleaning
3. Exploratory Analysis: Bivariate and Multivariate
4. Creating Visualisations to Support the Analysis
5. Deriving Insights and Stating Conclusions

---

**NOTE:** The marks given along with headings and sub-headings are cumulative marks for those particular headings/sub-headings.<br>

The actual marks for each task are specified within the tasks themselves.

For example, marks given with heading *2* or sub-heading *2.1* are the cumulative marks, for your reference only. <br>

The marks you will receive for completing tasks are given with the tasks.

Suppose the marks for two tasks are: 3 marks for 2.1.1 and 2 marks for 3.2.2, or
* 2.1.1 [3 marks]
* 3.2.2 [2 marks]

then, you will earn 3 marks for completing task 2.1.1 and 2 marks for completing task 3.2.2.


---

## Data Understanding
The yellow taxi trip records include fields capturing pick-up and drop-off dates/times, pick-up and drop-off locations, trip distances, itemized fares, rate types, payment types, and driver-reported passenger counts.

The data is stored in Parquet format (*.parquet*). The dataset is from 2009 to 2024. However, for this assignment, we will only be using the data from 2023.

The data for each month is present in a different parquet file. You will get twelve files for each of the months in 2023.

The data was collected and provided to the NYC Taxi and Limousine Commission (TLC) by technology providers like vendors and taxi hailing apps. <br>

You can find the link to the TLC trip records page here: https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page

###  Data Description
You can find the data description here: [Data Dictionary](https://www.nyc.gov/assets/tlc/downloads/pdf/data_dictionary_trip_records_yellow.pdf)

**Trip Records**



|Field Name       |description |
|:----------------|:-----------|
| VendorID | A code indicating the TPEP provider that provided the record. <br> 1= Creative Mobile Technologies, LLC; <br> 2= VeriFone Inc. |
| tpep_pickup_datetime | The date and time when the meter was engaged.  |
| tpep_dropoff_datetime | The date and time when the meter was disengaged.   |
| Passenger_count | The number of passengers in the vehicle. <br> This is a driver-entered value. |
| Trip_distance | The elapsed trip distance in miles reported by the taximeter. |
| PULocationID | TLC Taxi Zone in which the taximeter was engaged |
| DOLocationID | TLC Taxi Zone in which the taximeter was disengaged |
|RateCodeID |The final rate code in effect at the end of the trip.<br> 1 = Standard rate <br> 2 = JFK <br> 3 = Newark <br>4 = Nassau or Westchester <br>5 = Negotiated fare <br>6 = Group ride |
|Store_and_fwd_flag |This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server.  <br>Y= store and forward trip <br>N= not a store and forward trip |
|Payment_type| A numeric code signifying how the passenger paid for the trip. <br> 1 = Credit card <br>2 = Cash <br>3 = No charge <br>4 = Dispute <br>5 = Unknown <br>6 = Voided trip |
|Fare_amount| The time-and-distance fare calculated by the meter. <br>Extra Miscellaneous extras and surcharges.  Currently, this only includes the 0.50 and 1 USD rush hour and overnight charges. |
|MTA_tax |0.50 USD MTA tax that is automatically triggered based on the metered rate in use. |
|Improvement_surcharge | 0.30 USD improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015. |
|Tip_amount |Tip amount – This field is automatically populated for credit card tips. Cash tips are not included. |
| Tolls_amount | Total amount of all tolls paid in trip.  |
| total_amount | The total amount charged to passengers. Does not include cash tips. |
|Congestion_Surcharge |Total amount collected in trip for NYS congestion surcharge. |
| Airport_fee | 1.25 USD for pick up only at LaGuardia and John F. Kennedy Airports|

Although the amounts of extra charges and taxes applied are specified in the data dictionary, you will see that some cases have different values of these charges in the actual data.

**Taxi Zones**

Each of the trip records contains a field corresponding to the location of the pickup or drop-off of the trip, populated by numbers ranging from 1-263.

These numbers correspond to taxi zones, which may be downloaded as a table or map/shapefile and matched to the trip records using a join.

This is covered in more detail in later sections.

---

## **1** Data Preparation

<font color = red>[5 marks]</font> <br>

### Import Libraries

In [ ]:
# Import warnings

import warnings
warnings.filterwarnings('ignore')

In [4]:
# Import the libraries you will be using for analysis
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
# Recommended versions
# numpy version: 1.26.4
# pandas version: 2.2.2
# matplotlib version: 3.10.0
# seaborn version: 0.13.2

# Check versions
print("numpy version:", np.__version__)
print("pandas version:", pd.__version__)
print("matplotlib version:", plt.matplotlib.__version__)
print("seaborn version:", sns.__version__)

numpy version: 2.0.2
pandas version: 2.2.2
matplotlib version: 3.10.0
seaborn version: 0.13.2


### **1.1** Load the dataset
<font color = red>[5 marks]</font> <br>

You will see twelve files, one for each month.

To read parquet files with Pandas, you have to follow a similar syntax as that for CSV files.

`df = pd.read_parquet('file.parquet')`

In [115]:
# Try loading one file

# df = pd.read_parquet('2023-1.parquet')
# df.info()
df=pd.read_parquet('/content/sample_data/Trip_data/2023-1.parquet')
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3041714 entries, 0 to 3066765
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            floa

How many rows are there? Do you think handling such a large number of rows is computationally feasible when we have to combine the data for all twelve months into one?

To handle this, we need to sample a fraction of data from each of the files. How to go about that? Think of a way to select only some portion of the data from each month's file that accurately represents the trends.

#### Sampling the Data
> One way is to take a small percentage of entries for pickup in every hour of a date. So, for all the days in a month, we can iterate through the hours and select 5% values randomly from those. Use `tpep_pickup_datetime` for this. Separate date and hour from the datetime values and then for each date, select some fraction of trips for each of the 24 hours.

To sample data, you can use the `sample()` method. Follow this syntax:

```Python
# sampled_data is an empty DF to keep appending sampled data of each hour
# hour_data is the DF of entries for an hour 'X' on a date 'Y'

sample = hour_data.sample(frac = 0.05, random_state = 42)
# sample 0.05 of the hour_data
# random_state is just a seed for sampling, you can define it yourself

sampled_data = pd.concat([sampled_data, sample]) # adding data for this hour to the DF
```

This *sampled_data* will contain 5% values selected at random from each hour.

Note that the code given above is only the part that will be used for sampling and not the complete code required for sampling and combining the data files.

Keep in mind that you sample by date AND hour, not just hour. (Why?)

---

**1.1.1** <font color = red>[5 marks]</font> <br>
Figure out how to sample and combine the files.

**Note:** It is not mandatory to use the method specified above. While sampling, you only need to make sure that your sampled data represents the overall data of all the months accurately.

In [ ]:
# Sample the data
# It is recommmended to not load all the files at once to avoid memory overload

In [ ]:
# from google.colab import drive
# drive.mount('/content/drive')
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Take a small percentage of entries from each hour of every date.
# Iterating through the monthly data:
#   read a month file -> day -> hour: append sampled data -> move to next hour -> move to next day after 24 hours -> move to next month file
# Create a single dataframe for the year combining all the monthly data

# Select the folder having data files
import os

# Select the folder having data files
os.chdir('/content/sample_data/Trip_data')

# Create a list of all the twelve files to read
file_list = os.listdir()

# initialise an empty dataframe
df = pd.DataFrame()


# iterate through the list of files and sample one by one:
for file_name in file_list:
  try:
    # file path for the current file
    file_path = os.path.join(os.getcwd(), file_name)

    # Reading the current file
    temp_data=pd.read_parquet(file_path)



    # We will store the sampled data for the current date in this df by appending the sampled data from each hour to this
    # After completing iteration through each date, we will append this data to the final dataframe.


    # Loop through dates and then loop through every hour of each date
    for hour in range(24):
      # Iterate through each hour of the selected date
      hour_data = temp_data[temp_data['tpep_pickup_datetime'].dt.hour == hour]

      # Sample 5% of the hourly data randomly
    sampled_hour = hour_data.sample(frac=0.05, random_state=42)

      # Append the sampled data to the sampled_data dataframe
    sampled_data = pd.concat([sampled_data, sampled_hour])

      # add data of this hour to the dataframe
      # Concatenate the sampled data of all the dates to a single dataframe
    df = pd.concat([df, sampled_data])# we initialised this empty DF earlier

  except Exception as e:
    print(f"Error reading file {file_name}: {e}")

After combining the data files into one DataFrame, convert the new DataFrame to a CSV or parquet file and store it to use directly.

Ideally, you can try keeping the total entries to around 250,000 to 300,000.

In [8]:
# Store the df in csv/parquet
# df.to_parquet('')
df.to_parquet('/content/sample_data/Trip_data/NYC_Taxi_sampled_data.parquet')

## **2** Data Cleaning
<font color = red>[30 marks]</font> <br>

Now we can load the new data directly.

In [116]:
# Load the new data file
df=pd.read_parquet('/content/sample_data/Trip_data/NYC_Taxi_sampled_data.parquet')


In [117]:
# df.head()
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
2373315         2  2023-11-21 23:07:22   2023-11-21 23:49:11              1.0   
907694          2  2023-11-08 23:35:26   2023-11-08 23:59:14              1.0   
788006          1  2023-11-07 23:53:47   2023-11-08 00:02:24              1.0   
3226760         1  2023-11-04 23:42:25   2023-11-04 23:57:52              NaN   
1154802         2  2023-11-10 23:55:12   2023-11-11 00:26:48              2.0   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
2373315          27.98         4.0                  N           132   
907694            3.73         1.0                  N           114   
788006            3.40         1.0                  N           132   
3226760           2.00         NaN               None           161   
1154802          15.30         1.0                  N           132   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
2373315           265             1        166.8   1.00      0.5       20.00   
907694            143             1         23.3   1.00      0.5        4.00   
788006            216             1         15.6   2.75      0.5        3.95   
3226760           234             0         14.9   1.00      0.5        3.98   
1154802           255             1         61.8   1.00      0.5       16.07   

         tolls_amount  improvement_surcharge  total_amount  \
2373315           0.0                    1.0        191.05   
907694            0.0                    1.0         32.30   
788006            0.0                    1.0         23.80   
3226760           0.0                    1.0         23.88   
1154802           0.0                    1.0         82.12   

         congestion_surcharge  Airport_fee  airport_fee  
2373315                   0.0         1.75          NaN  
907694                    2.5         0.00          NaN  
788006                    0.0         1.75          NaN  
3226760                   NaN          NaN          NaN  
1154802                   0.0         1.75          NaN

In [118]:
# df.info()
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 689383 entries, 2373315 to 1031834
Data columns (total 20 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               689383 non-null  int64         
 1   tpep_pickup_datetime   689383 non-null  datetime64[us]
 2   tpep_dropoff_datetime  689383 non-null  datetime64[us]
 3   passenger_count        660713 non-null  float64       
 4   trip_distance          689383 non-null  float64       
 5   RatecodeID             660713 non-null  float64       
 6   store_and_fwd_flag     660713 non-null  object        
 7   PULocationID           689383 non-null  int64         
 8   DOLocationID           689383 non-null  int64         
 9   payment_type           689383 non-null  int64         
 10  fare_amount            689383 non-null  float64       
 11  extra                  689383 non-null  float64       
 12  mta_tax                689383 non-null  fl

#### **2.1** Fixing Columns
<font color = red>[10 marks]</font> <br>

Fix/drop any columns as you seem necessary in the below sections

**2.1.1** <font color = red>[2 marks]</font> <br>

Fix the index and drop unnecessary columns

In [119]:
# Fix the index and drop any columns that are not needed
df.reset_index(inplace=True)


In [120]:
df['store_and_fwd_flag'].value_counts()

store_and_fwd_flag
N    657540
Y      3173
Name: count, dtype: int64

In [121]:
df['tolls_amount'].describe()

count    689383.000000
mean          0.557005
std           2.007596
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max          80.910000
Name: tolls_amount, dtype: float64

In [122]:
df['tolls_amount'].value_counts()

tolls_amount
0.00     634537
6.55      35084
6.94      15926
12.75       830
3.00        461
          ...  
26.69         3
13.79         3
80.91         3
13.19         3
31.69         3
Name: count, Length: 102, dtype: int64

In [123]:
df['mta_tax'].describe()

count    689383.000000
mean          0.495957
std           0.045114
min          -0.500000
25%           0.500000
50%           0.500000
75%           0.500000
max           0.800000
Name: mta_tax, dtype: float64

In [124]:
df['mta_tax'].value_counts()

mta_tax
 0.50    683803
 0.00      5486
 0.05        39
-0.50        33
 0.80        22
Name: count, dtype: int64

In [125]:
(df['tpep_pickup_datetime'].dt.year == 2023).value_counts()

tpep_pickup_datetime
True    689383
Name: count, dtype: int64

In [126]:
df.drop(columns=['index', 'store_and_fwd_flag'],inplace=True)

In [127]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-11-21 23:07:22   2023-11-21 23:49:11              1.0   
1         2  2023-11-08 23:35:26   2023-11-08 23:59:14              1.0   
2         1  2023-11-07 23:53:47   2023-11-08 00:02:24              1.0   
3         1  2023-11-04 23:42:25   2023-11-04 23:57:52              NaN   
4         2  2023-11-10 23:55:12   2023-11-11 00:26:48              2.0   

   trip_distance  RatecodeID  PULocationID  DOLocationID  payment_type  \
0          27.98         4.0           132           265             1   
1           3.73         1.0           114           143             1   
2           3.40         1.0           132           216             1   
3           2.00         NaN           161           234             0   
4          15.30         1.0           132           255             1   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0        166.8   1.00      0.5       20.00           0.0   
1         23.3   1.00      0.5        4.00           0.0   
2         15.6   2.75      0.5        3.95           0.0   
3         14.9   1.00      0.5        3.98           0.0   
4         61.8   1.00      0.5       16.07           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0        191.05                   0.0         1.75   
1                    1.0         32.30                   2.5         0.00   
2                    1.0         23.80                   0.0         1.75   
3                    1.0         23.88                   NaN          NaN   
4                    1.0         82.12                   0.0         1.75   

   airport_fee  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

**2.1.2** <font color = red>[3 marks]</font> <br>
There are two airport fee columns. This is possibly an error in naming columns. Let's see whether these can be combined into a single column.

In [128]:
df['Airport_fee'].value_counts()

Airport_fee
 0.00    539173
 1.75     66914
 1.25     15106
-1.75        25
Name: count, dtype: int64

In [129]:
df['Airport_fee'].isnull().sum()

np.int64(68165)

In [130]:
df['airport_fee'].isnull().sum()

np.int64(649888)

In [131]:
#df.method({col: value}, inplace=True)'
df.fillna({'airport_fee': 0}, inplace=True)
df.fillna({'Airport_fee': 0}, inplace=True)

In [132]:
df['Airport_fee'].isnull().sum()

np.int64(0)

In [133]:
df['airport_fee'].isnull().sum()

np.int64(0)

In [134]:
# Combine the two airport fee columns
df['Airport_Fee_Combined'] = df.apply(lambda row: row['Airport_fee']+row['airport_fee'], axis=1)


In [135]:
df['Airport_Fee_Combined'].value_counts()

Airport_Fee_Combined
 0.00    601768
 1.75     66914
 1.25     20676
-1.75        25
Name: count, dtype: int64

In [136]:
df.drop(columns=['Airport_fee', 'airport_fee'], inplace=True)

In [137]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-11-21 23:07:22   2023-11-21 23:49:11              1.0   
1         2  2023-11-08 23:35:26   2023-11-08 23:59:14              1.0   
2         1  2023-11-07 23:53:47   2023-11-08 00:02:24              1.0   
3         1  2023-11-04 23:42:25   2023-11-04 23:57:52              NaN   
4         2  2023-11-10 23:55:12   2023-11-11 00:26:48              2.0   

   trip_distance  RatecodeID  PULocationID  DOLocationID  payment_type  \
0          27.98         4.0           132           265             1   
1           3.73         1.0           114           143             1   
2           3.40         1.0           132           216             1   
3           2.00         NaN           161           234             0   
4          15.30         1.0           132           255             1   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0        166.8   1.00      0.5       20.00           0.0   
1         23.3   1.00      0.5        4.00           0.0   
2         15.6   2.75      0.5        3.95           0.0   
3         14.9   1.00      0.5        3.98           0.0   
4         61.8   1.00      0.5       16.07           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  \
0                    1.0        191.05                   0.0   
1                    1.0         32.30                   2.5   
2                    1.0         23.80                   0.0   
3                    1.0         23.88                   NaN   
4                    1.0         82.12                   0.0   

   Airport_Fee_Combined  
0                  1.75  
1                  0.00  
2                  1.75  
3                  0.00  
4                  1.75

**2.1.3** <font color = red>[5 marks]</font> <br>
Fix columns with negative (monetary) values

In [139]:
# check where values of fare amount are negative
df.describe()


VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  689383.000000                      689383                      689383   
mean        1.762277  2023-07-02 02:37:44.302750  2023-07-02 02:54:08.517650   
min         1.000000         2023-01-01 23:00:44         2023-01-01 23:08:30   
25%         2.000000         2023-04-10 23:12:37         2023-04-10 23:24:59   
50%         2.000000         2023-06-12 23:56:06         2023-06-13 00:16:16   
75%         2.000000  2023-10-17 23:56:57.500000  2023-10-18 00:16:33.500000   
max         6.000000         2023-12-31 23:59:08         2024-01-01 00:44:46   
std         0.426914                         NaN                         NaN   

       passenger_count  trip_distance     RatecodeID   PULocationID  \
count    660713.000000  689383.000000  660713.000000  689383.000000   
mean          1.410914       4.046422       1.188348     156.184980   
min           0.000000       0.000000       1.000000       1.000000   
25%           1.000000       1.260000       1.000000     114.000000   
50%           1.000000       2.260000       1.000000     144.000000   
75%           2.000000       4.590000       1.000000     229.000000   
max           6.000000      71.100000      99.000000     265.000000   
std           0.900089       4.778010       3.256922      62.351322   

        DOLocationID   payment_type    fare_amount          extra  \
count  689383.000000  689383.000000  689383.000000  689383.000000   
mean      157.855493       1.129265      21.107195       1.729881   
min         1.000000       0.000000       0.000000       0.000000   
25%        93.000000       1.000000      10.000000       1.000000   
50%       161.000000       1.000000      14.900000       1.000000   
75%       233.000000       1.000000      24.700000       2.750000   
max       265.000000       4.000000     512.000000      10.250000   
std        73.872600       0.507642      18.733210       1.618286   

             mta_tax     tip_amount   tolls_amount  improvement_surcharge  \
count  689383.000000  689383.000000  689383.000000          689383.000000   
mean        0.495957       3.790306       0.557005               0.999315   
min        -0.500000       0.000000       0.000000              -1.000000   
25%         0.500000       1.000000       0.000000               1.000000   
50%         0.500000       3.000000       0.000000               1.000000   
75%         0.500000       4.855000       0.000000               1.000000   
max         0.800000     200.000000      80.910000               1.000000   
std         0.045114       4.140304       2.007596               0.026514   

        total_amount  congestion_surcharge  Airport_Fee_Combined  
count  689383.000000         660713.000000         689383.000000  
mean       30.595620              2.250728              0.207288  
min        -3.250000              0.000000             -1.750000  
25%        16.770000              2.500000              0.000000  
50%        22.250000              2.500000              0.000000  
75%        34.490000              2.500000              0.000000  
max       515.500000              2.500000              1.750000  
std        22.800673              0.749029              0.548874

Did you notice something different in the `RatecodeID` column for above records?

In [140]:
# Analyse RatecodeID for the negative fare amounts
df.RatecodeID.value_counts()


RatecodeID
1.0     625512
2.0      26917
5.0       4921
4.0       2093
99.0       719
3.0        551
Name: count, dtype: int64

In [141]:
df.RatecodeID.isnull().sum()

np.int64(28670)

In [142]:
df.fillna({'RatecodeID':1}, inplace=True)

In [143]:
df.RatecodeID.isnull().sum()

np.int64(0)

In [144]:
df['RatecodeID']=df['RatecodeID'].apply(lambda x: 6 if x == 99 else x)

In [145]:
df.RatecodeID.value_counts()

RatecodeID
1.0    654182
2.0     26917
5.0      4921
4.0      2093
6.0       719
3.0       551
Name: count, dtype: int64

In [146]:
df.congestion_surcharge.value_counts()

congestion_surcharge
2.5    594834
0.0     65879
Name: count, dtype: int64

In [147]:
df.congestion_surcharge.isnull().sum()

np.int64(28670)

In [148]:
df.fillna({'congestion_surcharge': 0}, inplace=True)

In [149]:
df.congestion_surcharge.isnull().sum()

np.int64(0)

In [150]:
df.congestion_surcharge.value_counts()

congestion_surcharge
2.5    594834
0.0     94549
Name: count, dtype: int64

In [151]:
# Find which columns have negative values
pd.DataFrame({
    'Column': df.columns.values,
    'Negative-Values': [df[col].min() if col in df.select_dtypes(include=[np.number]).columns else 0 for col in df.columns]
})


Column  Negative-Values
0                VendorID             1.00
1    tpep_pickup_datetime             0.00
2   tpep_dropoff_datetime             0.00
3         passenger_count             0.00
4           trip_distance             0.00
5              RatecodeID             1.00
6            PULocationID             1.00
7            DOLocationID             1.00
8            payment_type             0.00
9             fare_amount             0.00
10                  extra             0.00
11                mta_tax            -0.50
12             tip_amount             0.00
13           tolls_amount             0.00
14  improvement_surcharge            -1.00
15           total_amount            -3.25
16   congestion_surcharge             0.00
17   Airport_Fee_Combined            -1.75

In [ ]:
# columns with negative values


# improvement_surcharge
# total_amount
# Airport_Fee_Combined

In [152]:
df.head(3)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-11-21 23:07:22   2023-11-21 23:49:11              1.0   
1         2  2023-11-08 23:35:26   2023-11-08 23:59:14              1.0   
2         1  2023-11-07 23:53:47   2023-11-08 00:02:24              1.0   

   trip_distance  RatecodeID  PULocationID  DOLocationID  payment_type  \
0          27.98         4.0           132           265             1   
1           3.73         1.0           114           143             1   
2           3.40         1.0           132           216             1   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0        166.8   1.00      0.5       20.00           0.0   
1         23.3   1.00      0.5        4.00           0.0   
2         15.6   2.75      0.5        3.95           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  \
0                    1.0        191.05                   0.0   
1                    1.0         32.30                   2.5   
2                    1.0         23.80                   0.0   

   Airport_Fee_Combined  
0                  1.75  
1                  0.00  
2                  1.75

In [153]:
# fix these negative values
df['Airport_Fee_Combined'].value_counts()

Airport_Fee_Combined
 0.00    601768
 1.75     66914
 1.25     20676
-1.75        25
Name: count, dtype: int64

In [154]:
df['Airport_Fee_Combined']=df['Airport_Fee_Combined'].apply(lambda x: (x*-1) if x < 0 else x)

In [155]:
df['Airport_Fee_Combined'].value_counts()

Airport_Fee_Combined
0.00    601768
1.75     66939
1.25     20676
Name: count, dtype: int64

In [156]:
df['improvement_surcharge'].value_counts()

improvement_surcharge
 1.0    688868
 0.3       253
 0.0       229
-1.0        33
Name: count, dtype: int64

In [157]:
df['improvement_surcharge']=df['improvement_surcharge'].apply(lambda x: (x*-1) if x < 0 else x)

In [158]:
df['improvement_surcharge'].value_counts()

improvement_surcharge
1.0    688901
0.3       253
0.0       229
Name: count, dtype: int64

In [160]:
df['mta_tax'].value_counts()

mta_tax
 0.50    683803
 0.00      5486
 0.05        39
-0.50        33
 0.80        22
Name: count, dtype: int64

In [161]:
df['mta_tax']=df['mta_tax'].apply(lambda x: (x*-1) if x < 0 else x)

In [162]:
df['mta_tax'].value_counts()

mta_tax
0.50    683836
0.00      5486
0.05        39
0.80        22
Name: count, dtype: int64

In [163]:
df[df['total_amount']<0].value_counts()

VendorID  tpep_pickup_datetime  tpep_dropoff_datetime  passenger_count  trip_distance  RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  total_amount  congestion_surcharge  Airport_Fee_Combined
2         2023-04-18 23:00:44   2023-04-18 23:00:52    1.0              0.04           1.0         132           132           2             0.0          0.0    0.5      0.0         0.0           1.0                    -3.25         0.0                   1.75                    13
          2023-06-11 23:05:40   2023-06-11 23:29:44    1.0              11.12          4.0         132           265           2             0.0          0.0    0.5      0.0         0.0           1.0                    -3.25         0.0                   1.75                    12
          2023-12-13 23:08:31   2023-12-13 23:08:44    0.0              0.00           5.0         265           265           2             0.0          0.0    0.5      0.0         0.0           1.0                    -1.50         0.0                   0.00                     8
Name: count, dtype: int64

In [164]:
# it was observed that where total_amount is in negative value all values (fare_amount, extra,tip_amount,	tolls_amount) are zeros except mta_tax, Improvement_surcharge & Airport_fee
# total_amount made to zero

df['total_amount']=df['total_amount'].apply(lambda x: 0 if x < 0 else x)

In [165]:
df[df['total_amount']<0].value_counts()

Series([], Name: count, dtype: int64)

In [166]:
pd.DataFrame({
    'Column': df.columns.values,
    'Negative-Values': [df[col].min() if col in df.select_dtypes(include=[np.number]).columns else 0 for col in df.columns]
})


Column  Negative-Values
0                VendorID              1.0
1    tpep_pickup_datetime              0.0
2   tpep_dropoff_datetime              0.0
3         passenger_count              0.0
4           trip_distance              0.0
5              RatecodeID              1.0
6            PULocationID              1.0
7            DOLocationID              1.0
8            payment_type              0.0
9             fare_amount              0.0
10                  extra              0.0
11                mta_tax              0.0
12             tip_amount              0.0
13           tolls_amount              0.0
14  improvement_surcharge              0.0
15           total_amount              0.0
16   congestion_surcharge              0.0
17   Airport_Fee_Combined              0.0

### **2.2** Handling Missing Values
<font color = red>[10 marks]</font> <br>

**2.2.1**  <font color = red>[2 marks]</font> <br>
Find the proportion of missing values in each column




In [114]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 689383 entries, 0 to 689382
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               689383 non-null  int64         
 1   tpep_pickup_datetime   689383 non-null  datetime64[us]
 2   tpep_dropoff_datetime  689383 non-null  datetime64[us]
 3   passenger_count        660713 non-null  float64       
 4   trip_distance          689383 non-null  float64       
 5   RatecodeID             689383 non-null  float64       
 6   PULocationID           689383 non-null  int64         
 7   DOLocationID           689383 non-null  int64         
 8   payment_type           689383 non-null  int64         
 9   fare_amount            689383 non-null  float64       
 10  extra                  689383 non-null  float64       
 11  mta_tax                689383 non-null  float64       
 12  tip_amount             689383 non-null  floa

In [138]:
# Find the proportion of missing values in each column
df.isnull().sum()/len(df)*100

VendorID                 0.000000
tpep_pickup_datetime     0.000000
tpep_dropoff_datetime    0.000000
passenger_count          4.158791
trip_distance            0.000000
RatecodeID               4.158791
PULocationID             0.000000
DOLocationID             0.000000
payment_type             0.000000
fare_amount              0.000000
extra                    0.000000
mta_tax                  0.000000
tip_amount               0.000000
tolls_amount             0.000000
improvement_surcharge    0.000000
total_amount             0.000000
congestion_surcharge     4.158791
Airport_Fee_Combined     0.000000
dtype: float64

**2.2.2**  <font color = red>[3 marks]</font> <br>
Handling missing values in `passenger_count`

In [167]:
# Display the rows with null values
# Impute NaN values in 'passenger_count'
df.passenger_count.isnull().sum()

np.int64(28670)

In [168]:
df.passenger_count.value_counts()

passenger_count
1.0    479308
2.0    116654
3.0     26393
4.0     15662
5.0     10020
0.0      8065
6.0      4611
Name: count, dtype: int64

In [169]:
df.fillna({'passenger_count': 1}, inplace=True)

Did you find zeroes in passenger_count? Handle these.

In [170]:
df['passenger_count']=df['passenger_count'].apply(lambda x: 1 if x == 0 else x)

In [171]:
df.passenger_count.value_counts()

passenger_count
1.0    516043
2.0    116654
3.0     26393
4.0     15662
5.0     10020
6.0      4611
Name: count, dtype: int64

**2.2.3**  <font color = red>[2 marks]</font> <br>
Handle missing values in `RatecodeID`

In [ ]:
df.fillna({'RatecodeID':1}, inplace=True)

In [172]:
# Fix missing values in 'RatecodeID'
df['RatecodeID'].isnull().sum()

np.int64(0)

In [ ]:
df['RatecodeID']=df['RatecodeID'].apply(lambda x: 6 if x == 99 else x)

In [174]:
df['RatecodeID'].value_counts()

RatecodeID
1.0    654182
2.0     26917
5.0      4921
4.0      2093
6.0       719
3.0       551
Name: count, dtype: int64

**2.2.4**  <font color = red>[3 marks]</font> <br>
Impute NaN in `congestion_surcharge`

In [ ]:
df.fillna({'congestion_surcharge': 0}, inplace=True)

In [175]:
# handle null values in congestion_surcharge
df.congestion_surcharge.isnull().sum()



np.int64(0)

In [176]:
df.congestion_surcharge.value_counts()

congestion_surcharge
2.5    594834
0.0     94549
Name: count, dtype: int64

Are there missing values in other columns? Did you find NaN values in some other set of columns? Handle those missing values below.

In [177]:
# Handle any remaining missing values

df.isnull().sum()

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
Airport_Fee_Combined     0
dtype: int64

### **2.3** Handling Outliers
<font color = red>[10 marks]</font> <br>

Before we start fixing outliers, let's perform outlier analysis.

In [ ]:
# Describe the data and check if there are any potential outliers present
# Check for potential out of place values in various columns



**2.3.1**  <font color = red>[10 marks]</font> <br>
Based on the above analysis, it seems that some of the outliers are present due to errors in registering the trips. Fix the outliers.

Some points you can look for:
- Entries where `trip_distance` is nearly 0 and `fare_amount` is more than 300
- Entries where `trip_distance` and `fare_amount` are 0 but the pickup and dropoff zones are different (both distance and fare should not be zero for different zones)
- Entries where `trip_distance` is more than 250  miles.
- Entries where `payment_type` is 0 (there is no payment_type 0 defined in the data dictionary)

These are just some suggestions. You can handle outliers in any way you wish, using the insights from above outlier analysis.

How will you fix each of these values? Which ones will you drop and which ones will you replace?

First, let us remove 7+ passenger counts as there are very less instances.

In [ ]:
# remove passenger_count > 6


In [ ]:
# Continue with outlier handling



In [ ]:
# Do any columns need standardising?



## **3** Exploratory Data Analysis
<font color = red>[90 marks]</font> <br>

In [ ]:
df.columns.tolist()

#### **3.1** General EDA: Finding Patterns and Trends
<font color = red>[40 marks]</font> <br>

**3.1.1** <font color = red>[3 marks]</font> <br>
Categorise the varaibles into Numerical or Categorical.
* `VendorID`:
* `tpep_pickup_datetime`:
* `tpep_dropoff_datetime`:
* `passenger_count`:
* `trip_distance`:
* `RatecodeID`:
* `PULocationID`:
* `DOLocationID`:
* `payment_type`:
* `pickup_hour`:
* `trip_duration`:


The following monetary parameters belong in the same category, is it categorical or numerical?


* `fare_amount`
* `extra`
* `mta_tax`
* `tip_amount`
* `tolls_amount`
* `improvement_surcharge`
* `total_amount`
* `congestion_surcharge`
* `airport_fee`

##### Temporal Analysis

**3.1.2** <font color = red>[5 marks]</font> <br>
Analyse the distribution of taxi pickups by hours, days of the week, and months.

In [ ]:
# Find and show the hourly trends in taxi pickups



In [ ]:
# Find and show the daily trends in taxi pickups (days of the week)



In [ ]:
# Show the monthly trends in pickups



##### Financial Analysis

Take a look at the financial parameters like `fare_amount`, `tip_amount`, `total_amount`, and also `trip_distance`. Do these contain zero/negative values?

In [ ]:
# Analyse the above parameters



Do you think it is beneficial to create a copy DataFrame leaving out the zero values from these?

**3.1.3** <font color = red>[2 marks]</font> <br>
Filter out the zero values from the above columns.

**Note:** The distance might be 0 in cases where pickup and drop is in the same zone. Do you think it is suitable to drop such cases of zero distance?

In [ ]:
# Create a df with non zero entries for the selected parameters.



**3.1.4** <font color = red>[3 marks]</font> <br>
Analyse the monthly revenue (`total_amount`) trend

In [ ]:
# Group data by month and analyse monthly revenue



**3.1.5** <font color = red>[3 marks]</font> <br>
Show the proportion of each quarter of the year in the revenue

In [ ]:
# Calculate proportion of each quarter



**3.1.6** <font color = red>[3 marks]</font> <br>
Visualise the relationship between `trip_distance` and `fare_amount`. Also find the correlation value for these two.

**Hint:** You can leave out the trips with trip_distance = 0

In [ ]:
# Show how trip fare is affected by distance



**3.1.7** <font color = red>[5 marks]</font> <br>
Find and visualise the correlation between:
1. `fare_amount` and trip duration (pickup time to dropoff time)
2. `fare_amount` and `passenger_count`
3. `tip_amount` and `trip_distance`

In [ ]:
# Show relationship between fare and trip duration



In [ ]:
# Show relationship between fare and number of passengers



In [ ]:
# Show relationship between tip and trip distance



**3.1.8** <font color = red>[3 marks]</font> <br>
Analyse the distribution of different payment types (`payment_type`)

In [ ]:
# Analyse the distribution of different payment types (payment_type).




- 1= Credit card
- 2= Cash
- 3= No charge
- 4= Dispute



##### Geographical Analysis

For this, you have to use the *taxi_zones.shp* file from the *taxi_zones* folder.

There would be multiple files inside the folder (such as *.shx, .sbx, .sbn* etc). You do not need to import/read any of the files other than the shapefile, *taxi_zones.shp*.

Do not change any folder structure - all the files need to be present inside the folder for it to work.

The folder structure should look like this:
```
Taxi Zones
|- taxi_zones.shp.xml
|- taxi_zones.prj
|- taxi_zones.sbn
|- taxi_zones.shp
|- taxi_zones.dbf
|- taxi_zones.shx
|- taxi_zones.sbx

 ```

 You only need to read the `taxi_zones.shp` file. The *shp* file will utilise the other files by itself.

We will use the *GeoPandas* library for geopgraphical analysis
```
import geopandas as gpd
```

More about geopandas and shapefiles: [About](https://geopandas.org/en/stable/about.html)


Reading the shapefile is very similar to *Pandas*. Use `gpd.read_file()` function to load the data (*taxi_zones.shp*) as a GeoDataFrame. Documentation: [Reading and Writing Files](https://geopandas.org/en/stable/docs/user_guide/io.html)

In [ ]:
# !pip install geopandas

**3.1.9** <font color = red>[2 marks]</font> <br>
Load the shapefile and display it.

In [ ]:
# import geopandas as gpd


# Read the shapefile using geopandas
zones = # read the .shp file using gpd
zones.head()

Now, if you look at the DataFrame created, you will see columns like: `OBJECTID`,`Shape_Leng`, `Shape_Area`, `zone`, `LocationID`, `borough`, `geometry`.
<br><br>

Now, the `locationID` here is also what we are using to mark pickup and drop zones in the trip records.

The geometric parameters like shape length, shape area and geometry are used to plot the zones on a map.

This can be easily done using the `plot()` method.

In [ ]:
# print(zones.info())
# zones.plot()

Now, you have to merge the trip records and zones data using the location IDs.



**3.1.10** <font color = red>[3 marks]</font> <br>
Merge the zones data into trip data using the `locationID` and `PULocationID` columns.

In [ ]:
# Merge zones and trip records using locationID and PULocationID



**3.1.11** <font color = red>[3 marks]</font> <br>
Group data by location IDs to find the total number of trips per location ID

In [ ]:
# Group data by location and calculate the number of trips



**3.1.12** <font color = red>[2 marks]</font> <br>
Now, use the grouped data to add number of trips to the GeoDataFrame.

We will use this to plot a map of zones showing total trips per zone.

In [ ]:
# Merge trip counts back to the zones GeoDataFrame




The next step is creating a color map (choropleth map) showing zones by the number of trips taken.

Again, you can use the `zones.plot()` method for this. [Plot Method GPD](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.plot.html#geopandas.GeoDataFrame.plot)

But first, you need to define the figure and axis for the plot.

`fig, ax = plt.subplots(1, 1, figsize = (12, 10))`

This function creates a figure (fig) and a single subplot (ax)

---

After setting up the figure and axis, we can proceed to plot the GeoDataFrame on this axis. This is done in the next step where we use the plot method of the GeoDataFrame.

You can define the following parameters in the `zones.plot()` method:
```
column = '',
ax = ax,
legend = True,
legend_kwds = {'label': "label", 'orientation': "<horizontal/vertical>"}
```

To display the plot, use `plt.show()`.

**3.1.13** <font color = red>[3 marks]</font> <br>
Plot a color-coded map showing zone-wise trips

In [ ]:
# Define figure and axis


# Plot the map and display it



In [ ]:
# can you try displaying the zones DF sorted by the number of trips?



Here we have completed the temporal, financial and geographical analysis on the trip records.

**Compile your findings from general analysis below:**

You can consider the following points:

* Busiest hours, days and months
* Trends in revenue collected
* Trends in quarterly revenue
* How fare depends on trip distance, trip duration and passenger counts
* How tip amount depends on trip distance
* Busiest zones


#### **3.2** Detailed EDA: Insights and Strategies
<font color = red>[50 marks]</font> <br>

Having performed basic analyses for finding trends and patterns, we will now move on to some detailed analysis focussed on operational efficiency, pricing strategies, and customer experience.

##### Operational Efficiency

Analyze variations by time of day and location to identify bottlenecks or inefficiencies in routes

**3.2.1** <font color = red>[3 marks]</font> <br>
Identify slow routes by calculating the average time taken by cabs to get from one zone to another at different hours of the day.

Speed on a route *X* for hour *Y* = (*distance of the route X / average trip duration for hour Y*)

In [ ]:
# Find routes which have the slowest speeds at different times of the day



How does identifying high-traffic, high-demand routes help us?

**3.2.2** <font color = red>[3 marks]</font> <br>
Calculate the number of trips at each hour of the day and visualise them. Find the busiest hour and show the number of trips for that hour.

In [ ]:
# Visualise the number of trips per hour and find the busiest hour



Remember, we took a fraction of trips. To find the actual number, you have to scale the number up by the sampling ratio.

**3.2.3** <font color = red>[2 mark]</font> <br>
Find the actual number of trips in the five busiest hours

In [ ]:
# Scale up the number of trips

# Fill in the value of your sampling fraction and use that to scale up the numbers
sample_fraction =



**3.2.4** <font color = red>[3 marks]</font> <br>
Compare hourly traffic pattern on weekdays. Also compare for weekend.

In [ ]:
# Compare traffic trends for the week days and weekends



What can you infer from the above patterns? How will finding busy and quiet hours for each day help us?

**3.2.5** <font color = red>[3 marks]</font> <br>
Identify top 10 zones with high hourly pickups. Do the same for hourly dropoffs. Show pickup and dropoff trends in these zones.

In [ ]:
# Find top 10 pickup and dropoff zones



**3.2.6** <font color = red>[3 marks]</font> <br>
Find the ratio of pickups and dropoffs in each zone. Display the 10 highest (pickup/drop) and 10 lowest (pickup/drop) ratios.

In [ ]:
# Find the top 10 and bottom 10 pickup/dropoff ratios



**3.2.7** <font color = red>[3 marks]</font> <br>
Identify zones with high pickup and dropoff traffic during night hours (11PM to 5AM)

In [ ]:
# During night hours (11pm to 5am) find the top 10 pickup and dropoff zones
# Note that the top zones should be of night hours and not the overall top zones



Now, let us find the revenue share for the night time hours and the day time hours. After this, we will move to deciding a pricing strategy.

**3.2.8** <font color = red>[2 marks]</font> <br>
Find the revenue share for nighttime and daytime hours.

In [ ]:
# Filter for night hours (11 PM to 5 AM)



##### Pricing Strategy

**3.2.9** <font color = red>[2 marks]</font> <br>
For the different passenger counts, find the average fare per mile per passenger.

For instance, suppose the average fare per mile for trips with 3 passengers is 3 USD/mile, then the fare per mile per passenger will be 1 USD/mile.

In [ ]:
# Analyse the fare per mile per passenger for different passenger counts




**3.2.10** <font color = red>[3 marks]</font> <br>
Find the average fare per mile by hours of the day and by days of the week

In [ ]:
# Compare the average fare per mile for different days and for different times of the day



**3.2.11** <font color = red>[3 marks]</font> <br>
Analyse the average fare per mile for the different vendors for different hours of the day

In [ ]:
# Compare fare per mile for different vendors



**3.2.12** <font color = red>[5 marks]</font> <br>
Compare the fare rates of the different vendors in a tiered fashion. Analyse the average fare per mile for distances upto 2 miles. Analyse the fare per mile for distances from 2 to 5 miles. And then for distances more than 5 miles.


In [ ]:
# Defining distance tiers



##### Customer Experience and Other Factors

**3.2.13** <font color = red>[5 marks]</font> <br>
Analyse average tip percentages based on trip distances, passenger counts and time of pickup. What factors lead to low tip percentages?

In [ ]:
#  Analyze tip percentages based on distances, passenger counts and pickup times



Additional analysis [optional]: Let's try comparing cases of low tips with cases of high tips to find out if we find a clear aspect that drives up the tipping behaviours

In [ ]:
# Compare trips with tip percentage < 10% to trips with tip percentage > 25%



**3.2.14** <font color = red>[3 marks]</font> <br>
Analyse the variation of passenger count across hours and days of the week.

In [ ]:
# See how passenger count varies across hours and days




**3.2.15** <font color = red>[2 marks]</font> <br>
Analyse the variation of passenger counts across zones

In [ ]:
# How does passenger count vary across zones



In [ ]:
# For a more detailed analysis, we can use the zones_with_trips GeoDataFrame
# Create a new column for the average passenger count in each zone.



Find out how often surcharges/extra charges are applied to understand their prevalance

**3.2.16** <font color = red>[5 marks]</font> <br>
Analyse the pickup/dropoff zones or times when extra charges are applied more frequently

In [ ]:
# How often is each surcharge applied?



## **4** Conclusion
<font color = red>[15 marks]</font> <br>

### **4.1** Final Insights and Recommendations
<font color = red>[15 marks]</font> <br>

Conclude your analyses here. Include all the outcomes you found based on the analysis.

Based on the insights, frame a concluding story explaining suitable parameters such as location, time of the day, day of the week etc. to be kept in mind while devising a strategy to meet customer demand and optimise supply.

**4.1.1** <font color = red>[5 marks]</font> <br>
Recommendations to optimize routing and dispatching based on demand patterns and operational inefficiencies

**4.1.2** <font color = red>[5 marks]</font> <br>

Suggestions on strategically positioning cabs across different zones to make best use of insights uncovered by analysing trip trends across time, days and months.

**4.1.3** <font color = red>[5 marks]</font> <br>
Propose data-driven adjustments to the pricing strategy to maximize revenue while maintaining competitive rates with other vendors.